### PR TITLE
Static provisioning support for spectrocloud_cluster_aws

### PR DESCRIPTION
**Pull request by SivaanandM**
### Jira
https://spectrocloud.atlassian.net/browse/PLT-139

### Change Description
Terrafrom - Implementation for static provisioning support for `spectrocloud_cluster_aws` resource, Added support for setting `vpc_id` in the cloud_config and setting up `az_subnets` in the machine pool configs.

### Type of change
[*] Improvement in existing resource, added support for static provisioning in spectrocloud_cluster_aws.

### Documentation
API Reference - https://docs.spectrocloud.com/api/v1/clusters
Doc -  https://docs.spectrocloud.com/clusters/public-cloud/aws

### Validation

- Provisioned AWS cluster with VPC_ID and subnet configuration.
- Updated cluster configuration and scaled up and re-provisioned the cluster.
- Decommissioned the cluster.
- Provisioned AWS cluster without VPC_ID and subnet configuration, non static provision.

### Sample TF Configuration

```
resource "spectrocloud_cluster_aws" "cluster" {
  name             = var.cluster_name
  tags             = ["dev", "department:devops", "owner:bob"]
  cloud_account_id = data.spectrocloud_cloudaccount_aws.account.id

  cloud_config {
    ssh_key_name = "spectro22"
    region       = "us-east-2"
    vpc_id = "vpc-0d2a3f1799e4d6065"
  }

  cluster_profile {
    id = data.spectrocloud_cluster_profile.profile.id
  }

  backup_policy {
    schedule                  = "0 0 * * SUN"
    backup_location_id        = data.spectrocloud_backup_storage_location.bsl.id
    prefix                    = "prod-backup"
    expiry_in_hour            = 7200
    include_disks             = true
    include_cluster_resources = true
  }

  scan_policy {
    configuration_scan_schedule = "0 0 * * SUN"
    penetration_scan_schedule   = "0 0 * * SUN"
    conformance_scan_schedule   = "0 0 1 * *"
  }

  machine_pool {
    control_plane           = true
    control_plane_as_worker = true
    name                    = "master-pool"
    count                   = 1
    instance_type           = "t3.large"
    disk_size_gb            = 62
    azs                     = ["us-east-2a"]
    az_subnet {
        id = "subnet-08864975df862eb58"
        az = "us-east-2a"
      }
    az_subnet {
        id = "subnet-031a7ff4ff5e7fb9a"
        az = "us-east-2a"
      }
  }

  machine_pool {
    name          = "worker-basic"
    count         = 1
    instance_type = "t3.large"
    azs           = ["us-east-2a"]
    az_subnet {
        id = "subnet-08864975df862eb58"
        az = "us-east-2a"
      }
  }

}
```

### Support - 

Will update the documentation after review

### CODE REVIEW

File: examples/resources/spectrocloud_cluster_aws/resource.tf [hunk 1]
```diff
@@ -19,8 +19,9 @@ resource "spectrocloud_cluster_aws" "cluster" {
   cloud_account_id = data.spectrocloud_cloudaccount_aws.account.id
 
   cloud_config {
-    ssh_key_name = "default"
-    region       = "us-west-2"
+    ssh_key_name = "spectro22"
+    region       = "us-east-2"
+    vpc_id = "vpc-0d2a3f1799e4d6065"
```
Comment (kreeuwijk) [inline examples/resources/spectrocloud_cluster_aws/resource.tf:24]: please add spaces to line up with the other items:
```
vpc_id       = "vpc-0d2a3f1799e4d6065"
```

File: examples/resources/spectrocloud_cluster_aws/resource.tf [hunk 2]
```diff
@@ -69,14 +70,26 @@ resource "spectrocloud_cluster_aws" "cluster" {
     count                   = 1
     instance_type           = "t3.large"
     disk_size_gb            = 62
-    azs                     = ["us-west-2a"]
+    azs                     = ["us-east-2a"]
```
Comment (kreeuwijk) [inline examples/resources/spectrocloud_cluster_aws/resource.tf:73]: `azs` and `az_subnets` are mutually exclusive. An example should have `azs` for dynamic provisioning and `az_subnets` for static provisioning

File: examples/resources/spectrocloud_cluster_aws/resource.tf [hunk 3]
```diff
@@ -69,14 +70,26 @@ resource "spectrocloud_cluster_aws" "cluster" {
...
+        az = "us-east-2a"
+      }
   }
 
   machine_pool {
     name          = "worker-basic"
     count         = 1
     instance_type = "t3.large"
-    azs           = ["us-west-2a"]
+    azs           = ["us-east-2a"]
```
Comment (kreeuwijk) [inline examples/resources/spectrocloud_cluster_aws/resource.tf:88]: `azs` and `az_subnets` are mutually exclusive. An example should have `azs` for dynamic provisioning and `az_subnets` for static provisioning

File: spectrocloud/resource_cluster_aws.go [hunk 4]
```diff
@@ -267,6 +271,22 @@ func resourceClusterAws() *schema.Resource {
 								Type: schema.TypeString,
 							},
 						},
+						"az_subnet": {
```
Comment (nikchern) [inline spectrocloud/resource_cluster_aws.go:274]: Let's consider here similar to EKS schema or refactor EKS to use new schema.

File: pkg/client/test/resource_cluster_aws_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		t.Logf("Machine pool for master and worker is not returned by func - FlattenMachinePoolConfigsAws")
+	} else {
+		for i, _ := range machinePools {
+			k := machinePools[i].(map[string]interface{})
+			if k["update_strategy"] != "RollingUpdateScaleOut" {
+				t.Fail()
+				t.Logf("Machine pool update strategy is not set to it default - RollingUpdateScaleOut")
+			}
+			if k["name"] == "master-pool" {
+				if k["control_plane_as_worker"].(bool) != machinePoolConfig[i].UseControlPlaneAsWorker ||
```
[the remark this answers is not in the document]
Comment (nikchern) [reply pkg/client/test/resource_cluster_aws_test.go:64]: https://github.com/spectrocloud/terraform-provider-spectrocloud/blob/c8b1d06e794f274d9d751b310e126367af005426/spectrocloud/resource_cluster_attachment_test.go

File: spectrocloud/resource_cluster_aws.go (hunk 4 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_aws.go:274]: We cannot use same as EKS, In EKS schema we cant have two subnet with same availability zones, as below examples, @nikchern we need to think about fixing eks schema as well

Eg: 

```
az_subnets = {
      "us-west-2a" = "subnet-0d4978ddbff16c"
      "us-west-2a" = "subnet-0d4978ddbff16c"
    }
```
or
```
az_subnets = {
      "us-west-2a" = "subnet-0d4978ddbff16c"
    }
    az_subnets = {
      "us-west-2a" = "subnet-0d4978ddbff16c"
    }
```

File: pkg/client/test/resource_cluster_aws_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SivaanandM) [reply pkg/client/test/resource_cluster_aws_test.go]: updated

File: spectrocloud/resource_cluster_aws.go (hunk 4 above)
Comment (nikchern) [reply spectrocloud/resource_cluster_aws.go:274]: @kreeuwijk as discussed with @karl-cardenas-coding we will go with a more verbose schema: 

```
 az_subnet {
      subnet_id = "subnet-0fd3677d9c41c2d82"
      az_id = "us-west-2a"
    }
```

Comment (karl-cardenas-coding) [reply spectrocloud/resource_cluster_aws.go:274]: With this approach, we can use a `dynamic` block 😄

**Review by nikchern — APPROVED** [review]

File: spectrocloud/resource_cluster_aws.go [hunk 6]
```diff
@@ -260,13 +264,29 @@ func resourceClusterAws() *schema.Resource {
...
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"az": {
+										Type:     schema.TypeString,
+										Required: true,
```
Comment (karl-cardenas-coding) [inline spectrocloud/resource_cluster_aws.go:281]:
```suggestion
										Required: true,
										Description: "The AWS availability zone the subnet resides in.",
```
added description.

File: spectrocloud/resource_cluster_aws.go [hunk 7]
```diff
@@ -260,13 +264,29 @@ func resourceClusterAws() *schema.Resource {
...
+									},
+									"subnet_id": {
+										Type:     schema.TypeString,
```
Comment (karl-cardenas-coding) [inline spectrocloud/resource_cluster_aws.go:284]:
```suggestion
										Type:     schema.TypeString,
										Description: "The VPC subnet identifier. ",
```

**Review by karl-cardenas-coding — COMMENTED** [review]
I added descriptions to the new resources

**Pull request merged**